src/Config/ConfigurationResolver.php

```diff
@@ -0,0 +1,401 @@
+        'exclude' => [],
+        'config' => null,
+    ];
+
+    private $finder;
+
+    /**
+     * @var Container
+     */
+    private $container;
+    /**
+     * @var bool
+     */
+    private $configFinderIsOverridden;
+    private $cwd;
+    private $specificRulesets;
+
+    /**
+     * ConfigurationResolver constructor.
+     * @param Container $container
+     * @param $cwd
+     * @param array $options
+     */
+    public function __construct(Container $container, $cwd, array $options)
+    {
+        $this->container = $container;
+        $this->cwd = $cwd;
+        $this->defaultConfig = new Config();
+
+        foreach ($options as $name => $value) {
+            $this->setOption($name, $value);
+        }
+    }
+
+    public function getReporter()
+    {
+        $reporterServiceName = "reporter.{$this->config->getReporter()}";
+        if ($this->options['ruleset-class-name']) {
+            $reporterServiceName = "reporter.{$this->options['reporter-service-name']}";
+        }
+
+        return $this->container[$reporterServiceName];
+    }
+
+    /**
+     * Set option that will be resolved.
+     *
+     * @param string $name
+     * @param mixed  $value
+     */
+    private function setOption($name, $value)
+    {
+        if (!\array_key_exists($name, $this->options)) {
+            throw new InvalidConfigurationException(sprintf('Unknown option name: "%s".', $name));
+        }
+
+        $this->options[$name] = $value;
+    }
+
+    public function getRuleset(string $file)
+    {
+        $rulesetClassName = $this->getSpecificRuleset($file);
+
+        if (null === $rulesetClassName) {
+            $rulesetClassName = $this->config->getRuleset();
+            if ($this->options['ruleset-class-name']) {
+                $rulesetClassName = $this->options['ruleset-class-name'];
+            }
+        }
+
+        if (!class_exists($rulesetClassName)) {
+            throw new \InvalidArgumentException(sprintf('Ruleset class %s does not exist', $rulesetClassName));
+        }
+
+        if (!is_subclass_of($rulesetClassName, RulesetInterface::class)) {
+            throw new \InvalidArgumentException('Ruleset class must implement '.RulesetInterface::class);
+        }
+
+        return new $rulesetClassName();
+    }
+
+    /**
+     * File/Glob specific Ruleset definition. cannot be set via cli
+     *
+     * @param string $file
+     * @return mixed|null
+     */
+    private function getSpecificRuleset(string $file)
+    {
+        if (null === $this->specificRulesets) {
+            $this->specificRulesets = $this->getConfig()->getSpecificRulesets();
+        }
+
+        $file = $this->normalizePath($file);
+        foreach ($this->specificRulesets as $pattern => $rulesetClassName) {
+            // TODO: fnmatch allow setting flags https://www.php.net/manual/en/function.fnmatch.php
+            $pattern = $this->normalizePath($pattern);
+            if ($file === $pattern || fnmatch($pattern, $file)) {
+
```

Comment:
That extra line should be removed in my opinion.

Reply:
same as for the other code style comment